src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+   *     Craig-Interpolation.
+   */
+  private Term getCVC5Interpolation(
+      ImmutableList<Collection<Term>> pFormAAsList, ImmutableList<Collection<Term>> pFormBAsList) {
+
+    // Respect Asserted Formulas not in the Interpolation Pairs
+
+    ImmutableList<Collection<Term>> combinedInterpols =
+        Stream.concat(pFormAAsList.stream(), pFormBAsList.stream())
+            .collect(ImmutableList.toImmutableList());
+
+    // checks, that no Assertions in the Assertion stack are left out. Can happen in Tree
+    // Interpolation, if pPartitionedFormulas does not contain every Set of Formulas from the
+    // Assertion stack.
+    Collection<Term> extraAssertions = getAssertedTermsNotInCollection(combinedInterpols);
+
+    // Uses a separate Solver instance to leave the original solver-context unmodified
+    Solver interpolationSolver = new Solver();
+
+    setSolverOptions(seed, solverOptions, interpolationSolver);
+
+    Term interpolantA = buildConjunctionOfFormulasOverList(pFormAAsList, interpolationSolver);
+
+    if (!extraAssertions.isEmpty()) {
+      Term extraAssert = buildConjunctionOfFormulas(extraAssertions, interpolationSolver);
+      interpolantA = interpolationSolver.mkTerm(Kind.AND, extraAssert, interpolantA);
+    }
+    Term interpolantB = buildConjunctionOfFormulasOverList(pFormBAsList, interpolationSolver);
+
+    interpolationSolver.assertFormula(interpolantA);
+
+    Term interpolant =
+        interpolationSolver.getInterpolant(interpolationSolver.mkTerm(Kind.NOT, interpolantB));
+
+    interpolationSolver.deletePointer();
+
+    return interpolant;
+  }
+
+  /**
+   * Returns Asserted Formulas (from the formula Stack), not present in the given collection of
+   * Terms.
+   *
+   * @param pCombinedInterpols asserted Formulas to invert the selection
+   * @return asserted Formulas not in collTerms, but in the formula Stack
+   */
+  private Collection<Term> getAssertedTermsNotInCollection(
+      ImmutableList<Collection<Term>> pCombinedInterpols) {
+    ImmutableList<Term> retTerms =
+        pCombinedInterpols.stream()
+            .flatMap(Collection::stream)
+            .collect(ImmutableList.toImmutableList());
+
+    Set<Term> filteredAssertedTerms =
+        assertedFormulas.stream()
+            .flatMap(c -> c.stream())
+            .filter(n -> !retTerms.contains(n))
+            .collect(ImmutableSet.toImmutableSet());
+    return filteredAssertedTerms;
+  }
+
+  /**
+   * Turns a List of Collections of Formulas to a Single Conjunction of the Formulas e.g.:
+   * [[A,B],[C]] -> A/\B/\C.
+   *
+   * @param formula List of Collections of formulas
+   * @param usingSolver the CVC5 Solver Instance to use
+   * @return concatenated Formulas with AND as CVC5 Term
+   */
+  private Term buildConjunctionOfFormulasOverList(
+      ImmutableList<Collection<Term>> formula, Solver usingSolver) {
+    Collection<Term> eachCollectionOfTermsConcat =
```

Comment:
Instead of conjuncting several conjunctions from several lists, it would be much easier to only perform one conjunction of a longer (flattened) list.

The code might look like:
```
return buildConjunctionOfFormulas(FluentIterable.concat(formula).toSet(), usingSolver);
```

Reply:
Good Point.